File: version_java/src/main/java/com/example/version_java/di/RepositoryModule.java
package com.example.version_java.di;


import com.example.version_java.data.repo.CriminalRepository;
import com.example.version_java.data.repo.CriminalRepositoryImpl;
import com.example.version_java.data.repo.KakaoRepository;
import com.example.version_java.data.repo.KakaoRepositoryImpl;
import com.example.version_java.data.source.local.CrimianlLocalDataSource;
import com.example.version_java.data.source.local.CrimianlLocalDataSourceImpl;
import com.example.version_java.data.source.remote.CriminalRemoteDataSource;
import com.example.version_java.data.source.remote.CriminalRemoteDataSourceImpl;
import com.example.version_java.data.source.remote.KakaoRemoteDataSource;
import com.example.version_java.data.source.remote.KakaoRemoteDataSourceImpl;

import javax.inject.Singleton;

import dagger.Binds;
import dagger.Module;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;

@Module
@InstallIn(SingletonComponent.class)
public abstract class RepositoryModule {

    @Singleton
    @Binds
    public abstract CriminalRepository bindCriminalRepository(
            CriminalRepositoryImpl criminalRepositoryImpl
    );

    @Singleton
    @Binds
    public abstract CriminalRemoteDataSource bindCriminalRemoteDataSource(
            CriminalRemoteDataSourceImpl criminalRemoteDataSourceImpl
    );

    @Singleton
    @Binds
    public abstract CrimianlLocalDataSource bindCrimianlLocalDataSource(
            CrimianlLocalDataSourceImpl crimianlLocalDataSourceImpl
    );

    @Singleton
    @Binds
    public abstract KakaoRepository bindKakaoRepository(
            KakaoRepositoryImpl kakaoRepositoryImpl
    );

    @Singleton
    @Binds
    public abstract KakaoRemoteDataSource bindKakaoRemoteDataSource(
            KakaoRemoteDataSourceImpl kakaoRemoteDataSourceImpl
    );

}

File: app/src/main/java/com/example/crimenotification/di/RepositoryModule.kt
package com.example.crimenotification.di


import com.example.crimenotification.data.repo.*
import com.example.crimenotification.data.source.local.CriminalLocalDataSource
import com.example.crimenotification.data.source.local.CriminalLocalDataSourceImpl
import com.example.crimenotification.data.source.remote.*
import dagger.Binds
import dagger.Module
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
abstract class RepositoryModule {

    @Singleton
    @Binds
    abstract fun bindCriminalRepository(criminalRepositoryImpl: CriminalRepositoryImpl): CriminalRepository

    @Singleton
    @Binds
    abstract fun bindCriminalRemoteDataSource(criminalRemoteDataSourceImpl: CriminalRemoteDataSourceImpl): CriminalRemoteDataSource

    @Singleton
    @Binds
    abstract fun bindCriminalLocalDataSource(criminalLocalDataSourceImpl: CriminalLocalDataSourceImpl): CriminalLocalDataSource

    @Singleton
    @Binds
    abstract fun bindKakaoRepository(kakaoRepositoryImpl: KakaoRepositoryImpl): KakaoRepository

    @Singleton
    @Binds
    abstract fun bindCKakaoRemoteDataSource(kakaoRemoteDataSourceImpl: KakaoRemoteDataSourceImpl): KakaoRemoteDataSource

    @Singleton
    @Binds
    abstract fun bindFirebaseRepository(firebaseRepositoryImpl: FirebaseRepositoryImpl): FirebaseRepository

    @Singleton
    @Binds
    abstract fun bindFirebaseRemoteDataSource(firebaseRemoteDataSourceImpl: FirebaseRemoteDataSourceImpl): FirebaseRemoteDataSource

}

